pilot/pkg/config/clusterregistry/secretcontroller.go

```diff
@@ -0,0 +1,158 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusterregistry
+
+import (
+	"strings"
+	"time"
+
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	// "k8s.io/client-go/tools/clientcmd"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	mcLabel = "istio/multiCluster"
+)
+
+// Controller is the controller implementation for Secret resources
+type Controller struct {
+	kubeclientset kubernetes.Interface
+	namespace     string
+	secretsSynced cache.InformerSynced
+	cs            *ClusterStore
+}
+
+// NewController returns a new secret controller
+func NewController(
+	kubeclientset kubernetes.Interface,
+	kubeInformerFactory kubeinformers.SharedInformerFactory,
+	namespace string,
+	cs *ClusterStore) *Controller {
+
+	secretsInformer := kubeInformerFactory.Core().V1().Secrets()
+
+	controller := &Controller{
+		kubeclientset: kubeclientset,
+		namespace:     namespace,
+		secretsSynced: secretsInformer.Informer().HasSynced,
+		cs:            cs,
+	}
+
+	log.Info("Setting up event handlers")
+	secretsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.secretAdd,
+		DeleteFunc: controller.secretDelete,
+	})
+
+	return controller
+}
+
+// Run starts the controller until it receves a message over stopCh
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer runtime.HandleCrash()
+
+	// Start the informer factories to begin populating the informer caches
+	log.Info("Starting Secrets controller")
+
+	// Wait for the caches to be synced before starting workers
+	log.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.secretsSynced); !ok {
+		log.Errorf("failed to wait for caches to sync")
+	} else {
+		<-stopCh
+	}
+	log.Info("Shutting down workers")
+}
+
+func checkSecret(s *corev1.Secret) bool {
+	lv, ok := s.Labels[mcLabel]
+	if !ok {
+		return false
+	}
+	if strings.ToLower(lv) != "true" {
+		return false
+	}
+
+	return ok
+}
+
+func addMemberCluster(s *corev1.Secret, cs *ClusterStore) {
+	cs.storeLock.Lock()
+	defer cs.storeLock.Unlock()
+	// Check if there is already a cluster member with the specified
+	if _, ok := cs.clientConfigs[s.ObjectMeta.Name]; !ok {
+		log.Infof("Adding new cluster member: %s", s.ObjectMeta.Name)
+		clientConfig, err := clientcmd.Load(s.Data[s.ObjectMeta.Name])
+		if err != nil {
+			log.Errorf("failed to load client config from secret %s in namespace %s with error: %v",
+				s.ObjectMeta.Name, s.ObjectMeta.Namespace, err)
+		}
+		cs.clientConfigs[s.ObjectMeta.Name] = *clientConfig
+	}
+	log.Infof("Number of clusters in the cluster store: %d", len(cs.clientConfigs))
```

Comment:
Per discussion on slack channel, clusters will be retained for some time. In that case, Need to create a Cluster object and add it to cs.clusters.

Reply:
@baodongli instead of creating objects which are not really used, would it not make more sense to start using info from clientConfig? What do you use k8s_cr.Cluster for?

Reply:
They are used for now, for example,  to get Cluster Name. Also there is a GetPilotClusters() call.

Reply:
the same info exists in clientConfigs, you can retrieve cluster name form there.
If I remember right GetPilotClusters() dumps as a return the whole cluster store. Cluster struct is pretty useless, and exists just to not break code dealing with configmap, as soon as configmap is gone, clusters will also be eliminated.

Reply:
well, if the intent is not to be intrusive ....

Reply:
The number of clusters is best in a monitored variable. I believe expvar are exported to prometheus - can you add one ?